File: java/facade_pattern/FacadePattern.java
class Parser {
    public void parseRelevanceData() {
        System.out.println("Parsing Relevance Data");
    }

    public void parseRankerData() {
        System.out.println("Parsing Ranker Data");
    }
}

class RedisClient {
    public void saveRankerToRedis() {
        System.out.println("Saved Ranker Data in Redis");
    }

    public void saveRelevanceToRedis() {
        System.out.println("Saved Relevance Data in Redis");
    }
}

class MongoClient {
    public void saveRankerToMongo() {
        System.out.println("Saved Ranker Data in Mongo");
    }

    public void saveRelevanceToMongo() {
        System.out.println("Saved Relevance Data in Mongo");
    }
}

class RelevanceFacade {
    private final Parser parser;
    private final RedisClient redisClient;
    private final MongoClient mongoClient;

    public RelevanceFacade(Parser parser, RedisClient redisClient, MongoClient mongoClient) {
        this.parser = parser;
        this.redisClient = redisClient;
        this.mongoClient = mongoClient;
    }

    public void saveRelevance() {
        parser.parseRelevanceData();
        redisClient.saveRelevanceToRedis();
        mongoClient.saveRelevanceToMongo();
    }
}

class RankerFacade {
    private final Parser parser;
    private final RedisClient redisClient;

    public RankerFacade(Parser parser, RedisClient redisClient) {
        this.parser = parser;
        this.redisClient = redisClient;
    }

    public void saveRanker() {
        parser.parseRankerData();
        redisClient.saveRankerToRedis();
    }
}

public class FacadePattern {
    public static void main(String[] args) {
        Parser parser = new Parser();
        RedisClient redisClient = new RedisClient();
        MongoClient mongoClient = new MongoClient();

        RelevanceFacade relevanceFacade = new RelevanceFacade(parser, redisClient, mongoClient);
        relevanceFacade.saveRelevance();

        RankerFacade rankerFacade = new RankerFacade(parser, redisClient);
        rankerFacade.saveRanker();
    }
}

File: go/facade_pattern.go
package main

import "fmt"

type Parser struct{}

func (p *Parser) ParseRelevanceData() {
	fmt.Println("Parsing Relevance Data")
}
func (p *Parser) ParseRankerData() {
	fmt.Println("Parsing Ranker Data")
}

type RedisClient struct{}

func (r *RedisClient) SaveRankerToRedis() {
	fmt.Println("Saved Ranker Data in Redis")
}
func (r *RedisClient) SaveRelevanceToRedis() {
	fmt.Println("Saved Relevance Data in Redis")
}

type MongoClient struct{}

func (r *MongoClient) SaveRankerToMongo() {
	fmt.Println("Saved Ranker Data in Mongo")
}
func (r *MongoClient) SaveRelevanceToMongo() {
	fmt.Println("Saved Relevance Data in Mongo")
}

type RelevanceFacade struct {
	parser      *Parser
	redisClient *RedisClient
	mongoClient *MongoClient
}

func NewRelevanceFacade(parser *Parser, redisClient *RedisClient, mongoClient *MongoClient) RelevanceFacade {
	return RelevanceFacade{
		parser:      parser,
		redisClient: redisClient,
		mongoClient: mongoClient,
	}
}
func (r RelevanceFacade) SaveRelevance() {
	r.parser.ParseRelevanceData()
	r.redisClient.SaveRelevanceToRedis()
	r.mongoClient.SaveRelevanceToMongo()
}

type RankerFacade struct {
	parser      *Parser
	redisClient *RedisClient
}

func NewRankerFacade(parser *Parser, redisClient *RedisClient) RankerFacade {
	return RankerFacade{
		parser:      parser,
		redisClient: redisClient,
	}
}
func (r RankerFacade) SaveRanker() {
	r.parser.ParseRankerData()
	r.redisClient.SaveRankerToRedis()
}

func main() {
	parser := &Parser{}
	redisClient := &RedisClient{}
	mongoClient := &MongoClient{}

	relevanceFacade := NewRelevanceFacade(parser, redisClient, mongoClient)
	relevanceFacade.SaveRelevance()

	rankerFacade := NewRankerFacade(parser, redisClient)
	rankerFacade.SaveRanker()
}
